Scheduler predicate to cap AWS EBS volumes mounted on nodes

DirectXMan12 opened the pull request:
According to Amazon (see http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html#linux-specific-volume-limits), AWS instances with more AWS 40 volumes can cause boot issues (and there is also a hard cap on EBS volumes that's larger).  At least one of these 40 is the root volumes, and the rest can be other attached EBS volumes.

This PR introduces a scheduler predicate that allows administrators to cap the number of EBS volumes that kubelet would try to mount onto a given node.  The predicate would reject pods as not fitting if they would cause the total number of EBS volumes on the node to go over this limit.

DirectXMan12 committed:
Scheduler predicate for capping node volume count

For certain volume types (e.g. AWS EBS or GCE PD), a limitted
number of such volumes can be attached to a given node.  This commit
introduces a predicate with allows cluster admins to cap
the maximum number of volumes matching a particular type attached to a
given node.

The volume type is configurable by passing a pair of filter functions,
and the maximum number of such volumes is configurable to allow node
admins to reserve a certain number of volumes for system use.

By default, the predicate is exposed as MaxEBSVolumeCount and
MaxGCEPDVolumeCount (for AWS ElasticBlocKStore and GCE PersistentDisk
volumes, respectively), each of which can be configured using the
`KUBE_MAX_PD_VOLS` environment variable.

Fixes #7835

DirectXMan12 commented:
The second commit in this PR allows administrators to customize the maximum number of pods as they wish using an environment variable.  Since administrators may wish to mount several volumes for system use, it is important that the number be customizable.

It would be better to have generically parameterizable scheduler predicates (as opposed to the current situation which seems to only allow arguments to two specific types of predicates).  However, generically parameterizable scheduler predicates probably deserve a separate design issue, so this provides a method for customization in the mean time.

DirectXMan12 commented:
It appears that volumes are detached separately from the actual object deletion, but this PR should reach eventual consistency (i.e. it is possible when using this predicate to have more than the max number of volumes attached to the node if pods are deleted and new pods are accepted before the volumes are cleaned up, but this should not cause issues much versus the trade-off of trying to get more real-time information).

DirectXMan12 commented:
CC @ncdc

ncdc commented:
cc @kubernetes/rh-cluster-infra @kubernetes/rh-scalability

pmorie commented:
@kubernetes/rh-storage

ncdc commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -145,6 +145,127 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
 	return true, nil
 }
 
+type MaxEBSVolumeCountChecker struct {
```
Please add Godoc for exported types/funcs

DirectXMan12 replied:
Added on the `New` function like the other scheduler predicate factory functions.

ncdc commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -145,6 +145,127 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
+	newPodNamespace := pod.Namespace
+
+	newEBSVolumes := make(map[string]bool)
+	for _, vol := range pod.Spec.Volumes {
```
Extract this to a func? Looks like you could use it below as well.

timothysc commented:
will dig in a second, but needs a doc update on 'docs/devel/scheduler_algorithm.md'

timothysc commented on plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go:
```diff
@@ -80,6 +85,27 @@ func defaultPredicates() sets.String {
+				// According to Amazon, attaching more than 40 volumes can cause boot failures
+				// TODO: allow for generically parameterized scheduler predicates, because this is a bit ugly
+				var maxEBSVols uint = 39
+				if rawMaxEBSVols := os.Getenv("KUBE_MAX_EBS_VOLS"); rawMaxEBSVols != "" {
```
I don't know how I feel about the env-var overload to set, as it's an anomaly.  Is there actually a legit use case where we would want to mock this?

timothysc commented on plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go:
```diff
@@ -80,6 +89,27 @@ func defaultPredicates() sets.String {
+				// According to Amazon, attaching more than 40 volumes can cause boot failures
+				// TODO: allow for generically parameterized scheduler predicates, because this is a bit ugly
+				var maxEBSVols uint = DefaultMaxEBSVolumes
+				if rawMaxEBSVols := os.Getenv("KUBE_MAX_EBS_VOLS"); rawMaxEBSVols != "" {
```
Is there a legit reason why you would over-ride this value?  Also the env-var seems pretty hacky.

ncdc replied:
We don't know if the OpenShift Operations team (or other organizations' cluster administrators) want or plan to attach a fixed number of volumes to nodes. This is a way to allow the admins to reserve a fixed number of volumes outside of the scope of the pods.

We really don't like this approach for the long term, but it seems like a decent short term compromise. The scheduler predicate configuration mechanism is not generic at the moment, and I'd like to open a discussion with the community about how it should be designed going forward, but I don't want that to hold up this PR if we can avoid it.

timothysc replied:
Should be outlined in a doc add then on how to override.

DirectXMan12 commented:
@aveshagarwal I could do that, but I wanted the separate error messages for negative vs parse error.

aveshagarwal commented:
@DirectXMan12 that is fine. but you might think about changing default-nondefault-default sequence to default-default-nondefault or nondefault-default-default for readability point of view IMO.

aveshagarwal commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -145,6 +145,105 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
+		} else if vol.PersistentVolumeClaim != nil {
+			pvcName := vol.PersistentVolumeClaim.ClaimName
+			if pvcName == "" {
+				return fmt.Errorf("PersistentVolumeClaim had no name: %q", pvcName)
```
Not sure if I am missing something or misunderstood, but why to return here and not continue with other remaining vols? Similarly for other errors below.

DirectXMan12 replied:
AFAIK, that's a bad state -- you shouldn't have a PVC-based volume that has no target PVC.

aveshagarwal replied:
For my understanding, does that means if there are n vols (as vol in volumes), if there is an issue with one of the vols, ignore all?

DirectXMan12 replied:
The logic here is: If the pod is incorrectly specified, abort.

However, I could see the argument either way -- I'm mostly ambivalent on the matter.

markturansky replied:
A pod cannot use the persistentVolumeClaim volume without a ClaimName.  It would fail validation.  It is certainly an error here if there is no claimName.

As I wrote elsewhere in this thread, you need to use claim's pvc.Spec.VolumeName to get the backing PV and inspect its PersistentVolumeSource for a non-nil EBS volume (or GCE, if this PR covers that volume, too).

timothysc commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -145,6 +145,105 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
+	for _, vol := range volumes {
+		if vol.AWSElasticBlockStore != nil {
+			ebsVolumes[vol.AWSElasticBlockStore.VolumeID] = true
+		} else if vol.PersistentVolumeClaim != nil {
```
This api logic should probably be looked over by @markturansky.. looks ok to me, but I'm not a PV expert.

DirectXMan12 replied:
the logic is more or less copied from one of the other predicates below.

markturansky replied:
To really know what the PVClaim is backed by, you'd need to query for the PV and inspect its PersistentVolumeSource just like is happening with the VolumeSource in the code above.

timothysc commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -145,6 +145,105 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
+				return fmt.Errorf("PersistentVolumeClaim had no name: %q", pvcName)
+			}
+			pvc, err := c.pvcInfo.GetPersistentVolumeClaimInfo(namespace, pvcName)
+			if err != nil {
```
given the number of checks, consider collapsing the statements with ;

timothysc commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -145,6 +145,105 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
+				return err
+			}
+
+			if pvc == nil {
```
does this ever occur?

DirectXMan12 replied:
/me grumbles about not having a `Result<Err, Val>` type in Go to make these kind of cases clearer when reading the interfaces.

It looks like it can't according to the only implementation of the interface, but interface doesn't make it clear.  The code here is more or less copied from the `VolumeZoneChecker` predicate, and whoever wrote that put in a check.  I'll remove it.

DirectXMan12 commented:
> @DirectXMan12 that is fine. but you might think about changing default-nondefault-default sequence to default-default-nondefault or nondefault-default-default for readability point of view IMO.

@aveshagarwal done.

timothysc commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -145,6 +145,105 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
+	}
+
+	// count unique volumes
+	existingEBSVolumes := make(map[string]bool)
```
Don't you need just a raw count, not unique count?  This op is an expensive predicate.

ncdc replied:
The PVs and PVCs are in a cache, so that helps some

DirectXMan12 replied:
From what I saw in the EBS code, if the volume is already mounted, it will use the same device, so it shouldn't count towards the limit.  Someone more intimately familiar with the EBS mounter should feel free to correct me on that, though.

git blame points me to @justinsb, perhaps?

ncdc replied:
We need to count actual EBS volumes, and it needs to be unique. So if you mount the same volume in 40 pods, that's really only 1 volume.

markturansky replied:
Confirmed.  The same initial mount to the host will be used, so it counts as 1 attachment of the volume.  The 40 subsequent bind mounts to 40 pods still counts as 1 volume/attachment.

timothysc commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -145,6 +145,105 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
+			return false, err
+		}
+	}
+	numExistingVolumes := uint(len(existingEBSVolumes))
```
I don't believe you need to unit convert, you could keep all types the same and len is guaranteed not be non-negative according to spec. https://golang.org/ref/spec#Length_and_capacity

DirectXMan12 replied:
Are you saying convert the function to take a int as well?  While it doesn't make much of a difference, I do like functions being clear as to what input is valid in their types, as much as possible.

timothysc commented:
/cc @justinsb as I'm sure he would be interested.

ncdc commented on docs/devel/scheduler_algorithm.md:
```diff
@@ -47,6 +47,7 @@ The purpose of filtering the nodes is to filter out the nodes that do not meet c
 - `PodFitsHost`: Filter out all nodes except the one specified in the PodSpec's NodeName field.
 - `PodSelectorMatches`: Check if the labels of the node match the labels specified in the Pod's `nodeSelector` field ([Here](../user-guide/node-selection/) is an example of how to use `nodeSelector` field).
 - `CheckNodeLabelPresence`: Check if all the specified labels exist on a node or not, regardless of the value.
+- `MaxEBSVolumeCount`: Ensure that the number off attached ElasticBlockStore volumes does not exceed a maximum value (by default, 39, since Amazon reccomends a maximum of 40).
```
recommends

ncdc replied:
s/number off/number of/

timothysc replied:
link reference + document how to override.

jsafrane replied:
At least GCE hs similar limitation of how many volumes they can attach to a node (I'm not sure about OpenShift). I think it would be better to have single generic command line option for all of them instead of individual  `MaxEBSVolumeCount`/`MaxPDVolumeCount` (`MaxCinderVolumeCount`).

timothysc commented on plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go:
```diff
@@ -80,6 +89,27 @@ func defaultPredicates() sets.String {
 		factory.RegisterFitPredicate("HostName", predicates.PodFitsHost),
+
+		// Fit is determined by whether or not there would be too many AWS EBS volumes attached to the node
+		factory.RegisterFitPredicateFactory(
```
What's missing here is selective provider defaults.  /cc @davidopp

DirectXMan12 replied:
Agreed.  I was thinking about that while I was writing this.

markturansky commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -145,6 +145,97 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
+				return err
+			}
+
+			if pv.Spec.AWSElasticBlockStore != nil {
```
this looks good.  this is the lookup i was referring to.  It's difficult to follow the disjointed code amidst the comments in the PR.

pmorie commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -145,6 +145,97 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
+	return c.predicate
+}
+
+func (c *MaxEBSVolumeCountChecker) filterEBSVolumes(volumes []api.Volume, namespace string, ebsVolumes map[string]bool) error {
```
You can transform this into a generic implementation by injecting a `func([]api.Volume, string) (map[string]bool, error)`, and then make the EBS checker inject this implementation.  There will likely be other people interested in using this predicate for their clouds.

jsafrane replied:
Yes, at least GCE supports only 16 attached disks.

davidopp commented:
Can you make this work on GCE too? See https://github.com/kubernetes/kubernetes/issues/7835#issuecomment-99580642

justinsb commented:
Thanks for this; we definitely weren't handling this particularly well before.  Currently the AWS provider also has a hard-coded list of mountpoints that means we can't even attach as many as AWS theoreticlaly allows - I opened #19920.  We should make sure that the list of mountpoints that the AWS cloudprovider will use is at least as big as the default value that this PR sets (though I think that is a separate issue/PR from this one)!

DirectXMan12 commented:
I've updated the actual predicate to take a pair of "filter functions" (in a struct) which filter valid Volumes and PVCs, so the predicate hypothetically should be usable with any volume type.  I've introduced predicates for EBS volumes and GCE PD volumes.  The customization for both predicates is done through the `KUBE_MAX_PD_VOLS` environment variable.

ncdc commented on docs/devel/scheduler_algorithm.md:
```diff
@@ -47,6 +47,8 @@ The purpose of filtering the nodes is to filter out the nodes that do not meet c
 - `PodFitsHost`: Filter out all nodes except the one specified in the PodSpec's NodeName field.
 - `PodSelectorMatches`: Check if the labels of the node match the labels specified in the Pod's `nodeSelector` field ([Here](../user-guide/node-selection/) is an example of how to use `nodeSelector` field).
 - `CheckNodeLabelPresence`: Check if all the specified labels exist on a node or not, regardless of the value.
+- `MaxEBSVolumeCount`: Ensure that the number off attached ElasticBlockStore volumes does not exceed a maximum value (by default, 39, since Amazon recommends a maximum of 40 -- see [Amazon's documentation](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html#linux-specific-volume-limits)).  The maximum value can be controlled by setting the `KUBE_MAX_PD_VOLS` environment variable ...
```
s/number off/number of/

DirectXMan12 replied:
whoops, thought I fixed that ;-)

ncdc commented on docs/devel/scheduler_algorithm.md:
```diff
@@ -47,6 +47,8 @@ The purpose of filtering the nodes is to filter out the nodes that do not meet c
 - `PodSelectorMatches`: Check if the labels of the node match the labels specified in the Pod's `nodeSelector` field ([Here](../user-guide/node-selection/) is an example of how to use `nodeSelector` field).
 - `CheckNodeLabelPresence`: Check if all the specified labels exist on a node or not, regardless of the value.
+- `MaxEBSVolumeCount`: Ensure that the number off attached ElasticBlockStore volumes does not exceed a maximum value (by default, 39, since Amazon recommends a maximum of 40 -- see [Amazon's documentation](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html#linux-specific-volume-limits)).  The maximum value can be controlled by setting the `KUBE_MAX_PD_VOLS` environment variable ...
+- `MaxGCEPDVolumeCount`: Ensure that the number off attached GCE PersistentDisk volumes does not exceed a maximum value (by default, 16, which is the maximum GCE allows -- see [GCE's documentation](https://cloud.google.com/compute/docs/disks/persistent-disks#limits_for_predefined_machine_types)).  The maximum value can be controlled by setting the `KUBE_MAX_PD_VOLS` environment variable.
```
number of

ncdc commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -145,6 +145,141 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
+	pvcInfo    PersistentVolumeClaimInfo
+}
+
+// Information for how to filter PD Volumes when checking PD Volume caps
```
PDVolumeFilter contains information ...

ncdc commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -145,6 +145,141 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
+	FilterPersistentVolume func(pv *api.PersistentVolume) (id string, relevant bool)
+}
+
+// MaxPDVolumeCountPredicate evaluates whether a pod can fit based on the number of volumes which match
```
NewMaxPDVolumeCountPredicate creates a new ...

DirectXMan12 replied:
I was going based on the style of `NewVolumeZonePredicate` (which puts the predicate documentation on the `New` function), but there doesn't quite seem to be a consistent style in this file (other predicates put the predicate documentation and the predicate function, and don't have any documentation on the new function, etc).  What is the desired style here (I'm somewhat loathe to introduce yet another style here).

ncdc replied:
The godoc convention is that you start your doc w/the name of the item in question.

pmorie replied:
@DirectXMan12 http://blog.golang.org/godoc-documenting-go-code

DirectXMan12 replied:
Yeah, I've seen that.  I just wasn't sure if that convention was supposed to be more important that the style existing in the file.  I'll make the change.

pmorie commented on docs/devel/scheduler_algorithm.md:
```diff
@@ -47,6 +47,8 @@ The purpose of filtering the nodes is to filter out the nodes that do not meet c
 - `PodFitsHost`: Filter out all nodes except the one specified in the PodSpec's NodeName field.
 - `PodSelectorMatches`: Check if the labels of the node match the labels specified in the Pod's `nodeSelector` field ([Here](../user-guide/node-selection/) is an example of how to use `nodeSelector` field).
 - `CheckNodeLabelPresence`: Check if all the specified labels exist on a node or not, regardless of the value.
+- `MaxEBSVolumeCount`: Ensure that the number of attached ElasticBlockStore volumes does not exceed a maximum value (by default, 39, since Amazon recommends a maximum of 40 -- see [Amazon's documentation](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html#linux-specific-volume-limits)).  The maximum value can be controlled by setting the `KUBE_MAX_PD_VOLS` environment variable.
```
Eventually there will likely be at least one cloud (AWS) that has at more than one PD type (EBS and EFS), so I would avoid giving names like `KUBE_MAX_PD_VOLS`.  How about `MAX_ATTACHED_GCEPD_VOLS` and so forth?

DirectXMan12 replied:
It was recommended above that I just have one variable to make things simpler (I'm somewhat ambivalent on the matter, however).

pmorie commented on plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go:
```diff
@@ -18,14 +18,41 @@ limitations under the License.
+
+// getMaxVols checks the max PD volumes environment variable, otherwise returning a default value
+func getMaxVols(defaultVal uint) uint {
+	if rawMaxVols := os.Getenv("KUBE_MAX_PD_VOLS"); rawMaxVols != "" {
```
I would inject this name, it is not an invariant that there will always be only one PD type.

pmorie commented on plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go:
```diff
@@ -80,6 +107,25 @@ func defaultPredicates() sets.String {
+		factory.RegisterFitPredicateFactory(
+			"MaxGCEPDVolumeCount",
+			func(args factory.PluginFactoryArgs) algorithm.FitPredicate {
+				// TODO: allow for generically parameterized scheduler predicates, because this is a bit ugly
```
Nit: famous last words before all the flesh is melted off of your skull by the ugliness required to make something generic

DirectXMan12 replied:
:fire:

davidopp commented:
Is this ready for review now?

DirectXMan12 commented:
@davidopp It should be.  The one open question from @pmorie seems to be a matter of opinion, as the opposite was suggested above.

pmorie commented:
@DirectXMan12 which question is that?

pmorie commented on plugin/pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -771,3 +793,224 @@ func TestServiceAffinity(t *testing.T) {
+		existingPods []*api.Pod
+		maxVols      uint
+		fits         bool
+		test         string
```
nit: in almost every other test like this, the field in this struct you call `test` is called `name`

DirectXMan12 replied:
The rest of the tests in this file use `test`, and mixing and matching in a single file seems like it would be more confusing...

pmorie commented on plugin/pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -771,3 +793,224 @@ func TestServiceAffinity(t *testing.T) {
+			existingPods: []*api.Pod{twoVolPod, oneVolPod},
+			maxVols:      4,
+			fits:         true,
+			test:         "fits when node capacity >= new pod's EBS volumes",
```
nit, put test names at top of these anonymous struct declarations

DirectXMan12 replied:
[see above]

DirectXMan12 commented:
@pmorie https://github.com/kubernetes/kubernetes/pull/19880#discussion_r50498487 (vs a comment which seems to have disappeared)

pmorie commented:
@DirectXMan12 I have a strong preference to avoid a name which makes it sound like there's a single kind of PD per cloud.  We have at least one example of this, since AWS has EBS and EFS.  Unless someone wants to present a counter-argument, I think it should change.  That said, I will listen to any and all counter-arguments.

DirectXMan12 commented:
@pmorie my initial counter argument (aside from the fact that you probably don't care about limiting EFS volume count since that's just NFS) is that if you had multiple disk types that attached at the VM level like EBS, you'd probably want to limit them together as one predicate (since they'd probably count towards the same limit).  That's just a guess though -- someone can feel free to refute what I said.

That being said, I don't particularly care one way or the other -- I just want to avoid a back and forth situation where different people have different requirements for the patch :wink:

pmorie commented:
@DirectXMan12 there are some limits associated with EFS, but unfortunately the page doesn't doc any limits on total number of mounts per host.

Anyway, you've convinced me no one cares about this but me, so I stand down.

davidopp commented:
Someone please let me know when this is ready to review.

timothysc commented:
@davidopp I think we've (red hat) nit'd ourselves quite a bit here, please help us stop ;-)

davidopp commented on docs/devel/scheduler_algorithm.md:
```diff
@@ -47,6 +47,8 @@ The purpose of filtering the nodes is to filter out the nodes that do not meet c
 - `PodSelectorMatches`: Check if the labels of the node match the labels specified in the Pod's `nodeSelector` field ([Here](../user-guide/node-selection/) is an example of how to use `nodeSelector` field).
 - `CheckNodeLabelPresence`: Check if all the specified labels exist on a node or not, regardless of the value.
+- `MaxEBSVolumeCount`: Ensure that the number of attached ElasticBlockStore volumes does not exceed a maximum value (by default, 39, since Amazon recommends a maximum of 40 -- see [Amazon's documentation](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html#linux-specific-volume-limits)).  The maximum value can be controlled by setting the `KUBE_MAX_PD_VOLS` environment variable.
+- `MaxGCEPDVolumeCount`: Ensure that the number of attached GCE PersistentDisk volumes does not exceed a maximum value (by default, 16, which is the maximum GCE allows -- see [GCE's documentation](https://cloud.google.com/compute/docs/disks/persistent-disks#limits_for_predefined_machine_types)).  The maximum value can be controlled by setting the `KUBE_MAX_PD_VOLS` environment variable.
```
OK I gotta ask -- why did you chose amazon_recommended_max-1 as the default for EBS, but google_recommended_max as the default for GCE?

davidopp replied:
Oh, I see the comment later -- please mention this thing about the system root volume in the MaxEBSVolumeCount description above, otherwise people may wonder the same thing I did.

DirectXMan12 replied:
ack

davidopp commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -145,6 +145,142 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
+}
+
+// PDVolumeFilter contains information on how to filter PD Volumes when checking PD Volume caps
+type PDVolumeFilter struct {
```
The name of this type is confusing because it contains a filter for normal volumes and one for persistent volumes. Would it be better to call it a VolumeFilter?

DirectXMan12 replied:
fair point, done.

davidopp commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -145,6 +145,142 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
+
+// NewMaxPDVolumeCountPredicate creates a predicate which evaluates whether a pod can fit based on the
+// number of volumes which match a filter that it requests, and those that are already present.  The
+// maximum number is configurable; Amazon recommends no more than 40 EBS volumes, while GCE reccomends
```
I would drop the part of this sentence after the semicolon because it's just more places you'll have to update when this number changes, and also the constant is defined in a different file.

DirectXMan12 replied:
ack, done.

davidopp commented on plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go:
```diff
@@ -18,14 +18,41 @@ limitations under the License.
+const DefaultMaxGCEPDVolumes = 16
+
+// getMaxVols checks the max PD volumes environment variable, otherwise returning a default value
+func getMaxVols(defaultVal uint) uint {
```
AIUI Go style is to avoid using "unit" except in very particular situations like bitmasks. I would suggest changing all of this to just regular int.

DirectXMan12 replied:
ack, done

davidopp commented:
Other than my minor comments this looks fine.

DirectXMan12 commented:
Minor nits should all be fixed.

davidopp commented on plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go:
```diff
@@ -80,6 +107,25 @@ func defaultPredicates() sets.String {
+			},
+		),
+
+		factory.RegisterFitPredicateFactory(
```
Add comment like you did for the other one: "// Fit is determined by whether or not there would be too many GCE PD volumes attached to the node"

DirectXMan12 replied:
ack

davidopp commented:
Please squash the commits, then I will LGTM.

DirectXMan12 commented:
squahsed

davidopp commented:
LGTM

hongchaodeng commented:
Please /cc @kubernetes/sig-scheduling for any scheduler changes. Thanks!
